File: domains/test/python/src/ingestion/schema.py
"""
Schema detection and validation for ingested data.

This module provides functionality to detect, validate, and enforce
schemas on ingested data.
"""

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, List, Optional, Union, Type
import json
import logging

from core.exceptions import SchemaError, ValidationError
from core.validators import BaseValidator, TypeValidator, SchemaValidator


class DataType(Enum):
    """Supported data types in schemas."""
    STRING = "string"
    INTEGER = "integer"
    FLOAT = "float"
    BOOLEAN = "boolean"
    DATETIME = "datetime"
    DATE = "date"
    TIME = "time"
    OBJECT = "object"
    ARRAY = "array"
    NULL = "null"
    UNKNOWN = "unknown"


@dataclass
class FieldSchema:
    """Schema definition for a single field."""
    
    name: str
    data_type: DataType
    nullable: bool = True
    unique: bool = False
    primary_key: bool = False
    foreign_key: Optional[str] = None
    default: Any = None
    min_value: Optional[Union[int, float]] = None
    max_value: Optional[Union[int, float]] = None
    min_length: Optional[int] = None
    max_length: Optional[int] = None
    pattern: Optional[str] = None
    enum_values: Optional[List[Any]] = None
    description: str = ""
    metadata: Dict[str, Any] = field(default_factory=dict)
    
    def validate_value(self, value: Any) -> bool:
        """Validate a value against this field schema.
        
        Args:
            value: Value to validate
            
        Returns:
            True if value is valid
            
        Raises:
            ValidationError: If value is invalid
        """
        # Check null values
        if value is None:
            if not self.nullable:
                raise ValidationError(f"Field '{self.name}' cannot be null")
            return True
        
        # Check data type
        if not self._check_type(value):
            raise ValidationError(
                f"Field '{self.name}' expected type {self.data_type.value}, "
                f"got {type(value).__name__}"
            )
        
        # Check numeric constraints
        if self.data_type in [DataType.INTEGER, DataType.FLOAT]:
            if self.min_value is not None and value < self.min_value:
                raise ValidationError(
                    f"Field '{self.name}' value {value} is below minimum {self.min_value}"
                )
            if self.max_value is not None and value > self.max_value:
                raise ValidationError(
                    f"Field '{self.name}' value {value} is above maximum {self.max_value}"
                )
        
        # Check string constraints
        if self.data_type == DataType.STRING:
            if self.min_length is not None and len(value) < self.min_length:
                raise ValidationError(
                    f"Field '{self.name}' length {len(value)} is below minimum {self.min_length}"
                )
            if self.max_length is not None and len(value) > self.max_length:
                raise ValidationError(
                    f"Field '{self.name}' length {len(value)} is above maximum {self.max_length}"
                )
            
            if self.pattern:
                import re
                if not re.match(self.pattern, value):
                    raise ValidationError(
                        f"Field '{self.name}' value does not match pattern {self.pattern}"
                    )
        
        # Check enum values
        if self.enum_values and value not in self.enum_values:
            raise ValidationError(
                f"Field '{self.name}' value {value} not in allowed values: {self.enum_values}"
            )
        
        return True
    
    def _check_type(self, value: Any) -> bool:
        """Check if value matches expected type.
        
        Args:
            value: Value to check
            
        Returns:
            True if type matches
        """
        type_checks = {
            DataType.STRING: lambda v: isinstance(v, str),
            DataType.INTEGER: lambda v: isinstance(v, int) and not isinstance(v, bool),
            DataType.FLOAT: lambda v: isinstance(v, (int, float)) and not isinstance(v, bool),
            DataType.BOOLEAN: lambda v: isinstance(v, bool),
            DataType.OBJECT: lambda v: isinstance(v, dict),
            DataType.ARRAY: lambda v: isinstance(v, list),
            DataType.DATETIME: lambda v: isinstance(v, str),  # Simplified check
            DataType.DATE: lambda v: isinstance(v, str),
            DataType.TIME: lambda v: isinstance(v, str),
        }
        
        checker = type_checks.get(self.data_type)
        return checker(value) if checker else True


@dataclass
class TableSchema:
    """Schema definition for a table/collection."""
    
    name: str
    fields: List[FieldSchema]
    primary_key: Optional[List[str]] = None
    foreign_keys: Optional[Dict[str, str]] = None
    indexes: Optional[List[List[str]]] = None
    constraints: Optional[List[str]] = None
    description: str = ""
    metadata: Dict[str, Any] = field(default_factory=dict)
    
    def __post_init__(self):
        """Post-initialization processing."""
        # Build field index for quick lookup
        self._field_index = {field.name: field for field in self.fields}
        
        # Auto-detect primary key if not specified
        if not self.primary_key:
            pk_fields = [f.name for f in self.fields if f.primary_key]
            if pk_fields:
                self.primary_key = pk_fields
    
    def get_field(self, name: str) -> Optional[FieldSchema]:
        """Get field schema by name.
        
        Args:
            name: Field name
            
        Returns:
            Field schema or None
        """
        return self._field_index.get(name)
    
    def validate_record(self, record: Dict[str, Any]) -> bool:
        """Validate a record against this schema.
        
        Args:
            record: Record to validate
            
        Returns:
            True if record is valid
            
        Raises:
            ValidationError: If record is invalid
        """
        # Check for required fields
        record_fields = set(record.keys())
        schema_fields = set(self._field_index.keys())
        
        # Check for missing required fields
        required_fields = {f.name for f in self.fields if not f.nullable and f.default is None}
        missing = required_fields - record_fields
        if missing:
            raise ValidationError(f"Missing required fields: {missing}")
        
        # Check for unexpected fields (optional, could be configurable)
        unexpected = record_fields - schema_fields
        if unexpected:
            logging.warning(f"Unexpected fields in record: {unexpected}")
        
        # Validate each field
        for field_name, value in record.items():
            if field_name in self._field_index:
                field_schema = self._field_index[field_name]
                field_schema.validate_value(value)
        
        return True
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert schema to dictionary representation.
        
        Returns:
            Schema as dictionary
        """
        return {
            'name': self.name,
            'fields': [
                {
                    'name': f.name,
                    'type': f.data_type.value,
                    'nullable': f.nullable,
                    'constraints': {
                        k: v for k, v in {
                            'unique': f.unique,
                            'primary_key': f.primary_key,
                            'foreign_key': f.foreign_key,
                            'min_value': f.min_value,
                            'max_value': f.max_value,
                            'min_length': f.min_length,
                            'max_length': f.max_length,
                            'pattern': f.pattern,
                            'enum': f.enum_values
                        }.items() if v
                    }
                }
                for f in self.fields
            ],
            'primary_key': self.primary_key,
            'foreign_keys': self.foreign_keys,
            'indexes': self.indexes
        }
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'TableSchema':
        """Create schema from dictionary.
        
        Args:
            data: Schema dictionary
            
        Returns:
            TableSchema instance
        """
        fields = []
        for field_data in data.get('fields', []):
            constraints = field_data.get('constraints', {})
            field = FieldSchema(
                name=field_data['name'],
                data_type=DataType(field_data['type']),
                nullable=field_data.get('nullable', True),
                unique=constraints.get('unique', False),
                primary_key=constraints.get('primary_key', False),
                foreign_key=constraints.get('foreign_key'),
                min_value=constraints.get('min_value'),
                max_value=constraints.get('max_value'),
                min_length=constraints.get('min_length'),
                max_length=constraints.get('max_length'),
                pattern=constraints.get('pattern'),
                enum_values=constraints.get('enum')
            )
            fields.append(field)
        
        return cls(
            name=data['name'],
            fields=fields,
            primary_key=data.get('primary_key'),
            foreign_keys=data.get('foreign_keys'),
            indexes=data.get('indexes')
        )


def detect_data_type(value: Any) -> DataType:
    """Detect the data type of a value.
    
    Args:
        value: Value to analyze
        
    Returns:
        Detected data type
    """
    if value is None:
        return DataType.NULL
    elif isinstance(value, bool):
        return DataType.BOOLEAN
    elif isinstance(value, int):
        return DataType.INTEGER
    elif isinstance(value, float):
        return DataType.FLOAT
    elif isinstance(value, str):
        # Try to detect datetime/date patterns
        import re
        if re.match(r'^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}', value):
            return DataType.DATETIME
        elif re.match(r'^\d{4}-\d{2}-\d{2}$', value):
            return DataType.DATE
        elif re.match(r'^\d{2}:\d{2}:\d{2}$', value):
            return DataType.TIME
        else:
            return DataType.STRING
    elif isinstance(value, dict):
        return DataType.OBJECT
    elif isinstance(value, list):
        return DataType.ARRAY
    else:
        return DataType.UNKNOWN


def detect_schema(records: List[Dict[str, Any]], 
                  sample_size: Optional[int] = None) -> Dict[str, Any]:
    """Detect schema from sample records.
    
    Args:
        records: Sample records
        sample_size: Maximum number of records to analyze
        
    Returns:
        Detected schema dictionary
    """
    if not records:
        return {'fields': [], 'record_count': 0}
    
    # Limit sample size if specified
    if sample_size:
        records = records[:sample_size]
    
    # Collect field information
    field_info = {}
    
    for record in records:
        for field_name, value in record.items():
            if field_name not in field_info:
                field_info[field_name] = {
                    'types': [],
                    'nullable': False,
                    'values': set(),
                    'min': None,
                    'max': None,
                    'min_length': None,
                    'max_length': None
                }
            
            info = field_info[field_name]
            
            # Track data types
            data_type = detect_data_type(value)
            info['types'].append(data_type)
            
            # Track null values
            if value is None:
                info['nullable'] = True
            else:
                # Track unique values (for potential enum detection)
                if len(info['values']) < 100:  # Limit to avoid memory issues
                    info['values'].add(value)
                
                # Track numeric ranges
                if isinstance(value, (int, float)):
                    if info['min'] is None or value < info['min']:
                        info['min'] = value
                    if info['max'] is None or value > info['max']:
                        info['max'] = value
                
                # Track string lengths
                if isinstance(value, str):
                    length = len(value)
                    if info['min_length'] is None or length < info['min_length']:
                        info['min_length'] = length
                    if info['max_length'] is None or length > info['max_length']:
                        info['max_length'] = length
    
    # Build schema from collected information
    fields = []
    
    for field_name, info in field_info.items():
        # Determine primary data type (most common non-null type)
        type_counts = {}
        for dt in info['types']:
            if dt != DataType.NULL:
                type_counts[dt] = type_counts.get(dt, 0) + 1
        
        if type_counts:
            primary_type = max(type_counts, key=type_counts.get)
        else:
            primary_type = DataType.STRING  # Default
        
        field_schema = {
            'name': field_name,
            'type': primary_type.value,
            'nullable': info['nullable']
        }
        
        # Add constraints if detected
        if info['values'] and len(info['values']) < 10:
            # Potential enum field
            field_schema['enum'] = list(info['values'])
        
        if primary_type in [DataType.INTEGER, DataType.FLOAT]:
            if info['min'] is not None:
                field_schema['min_value'] = info['min']
            if info['max'] is not None:
                field_schema['max_value'] = info['max']
        
        if primary_type == DataType.STRING:
            if info['min_length'] is not None:
                field_schema['min_length'] = info['min_length']
            if info['max_length'] is not None:
                field_schema['max_length'] = info['max_length']
        
        fields.append(field_schema)
    
    return {
        'fields': fields,
        'record_count': len(records),
        'detected': True
    }


class SchemaEvolution:
    """Handles schema evolution and compatibility."""
    
    def __init__(self):
        """Initialize schema evolution handler."""
        self.logger = logging.getLogger(__name__)
    
    def compare_schemas(self, old_schema: TableSchema, 
                       new_schema: TableSchema) -> Dict[str, Any]:
        """Compare two schemas and identify differences.
        
        Args:
            old_schema: Previous schema version
            new_schema: New schema version
            
        Returns:
            Dictionary describing differences
        """
        differences = {
            'compatible': True,
            'added_fields': [],
            'removed_fields': [],
            'modified_fields': [],
            'type_changes': []
        }
        
        old_fields = {f.name: f for f in old_schema.fields}
        new_fields = {f.name: f for f in new_schema.fields}
        
        # Find added fields
        for name in new_fields:
            if name not in old_fields:
                differences['added_fields'].append(name)
        
        # Find removed fields
        for name in old_fields:
            if name not in new_fields:
                differences['removed_fields'].append(name)
                # Removed required fields break compatibility
                if not old_fields[name].nullable:
                    differences['compatible'] = False
        
        # Find modified fields
        for name in old_fields:
            if name in new_fields:
                old_field = old_fields[name]
                new_field = new_fields[name]
                
                if old_field.data_type != new_field.data_type:
                    differences['type_changes'].append({
                        'field': name,
                        'old_type': old_field.data_type.value,
                        'new_type': new_field.data_type.value
                    })
                    # Type changes usually break compatibility
                    differences['compatible'] = False
                
                # Check other modifications
                modifications = []
                if old_field.nullable != new_field.nullable:
                    modifications.append('nullable')
                    if not new_field.nullable and old_field.nullable:
                        differences['compatible'] = False
                
                if modifications:
                    differences['modified_fields'].append({
                        'field': name,
                        'modifications': modifications
                    })
        
        return differences

File: domains/test/javascript/src/ingestion/schema.js
// ingestion/schema.js
/**
 * Schema detection and validation for the ingestion module.
 * 
 * This module provides utilities for detecting data schemas,
 * validating data against schemas, and schema transformation.
 */

import { SchemaError, ValidationError } from '../core/exceptions.js';
import { BaseValidator } from '../core/validators.js';

// Data types that can be detected
export const DataType = Object.freeze({
    STRING: 'string',
    INTEGER: 'integer',
    FLOAT: 'float',
    BOOLEAN: 'boolean',
    DATETIME: 'datetime',
    DATE: 'date',
    TIME: 'time',
    ARRAY: 'array',
    OBJECT: 'object',
    NULL: 'null',
    UNKNOWN: 'unknown'
});

/**
 * Represents a field in a data schema
 */
export class SchemaField {
    constructor(name, type, options = {}) {
        this.name = name;
        this.type = type;
        this.nullable = options.nullable ?? true;
        this.unique = options.unique || false;
        this.primaryKey = options.primaryKey || false;
        this.foreignKey = options.foreignKey || null;
        this.defaultValue = options.defaultValue || null;
        this.minValue = options.minValue || null;
        this.maxValue = options.maxValue || null;
        this.minLength = options.minLength || null;
        this.maxLength = options.maxLength || null;
        this.pattern = options.pattern || null;
        this.enum = options.enum || null;
        this.description = options.description || '';
    }

    validate(value) {
        // Check null values
        if (value === null || value === undefined) {
            if (!this.nullable) {
                throw new ValidationError(`Field '${this.name}' cannot be null`);
            }
            return true;
        }

        // Check type
        if (!this._validateType(value)) {
            throw new ValidationError(
                `Field '${this.name}' expected type ${this.type}, got ${typeof value}`
            );
        }

        // Check constraints
        this._validateConstraints(value);

        return true;
    }

    _validateType(value) {
        switch (this.type) {
            case DataType.STRING:
                return typeof value === 'string';
            case DataType.INTEGER:
                return Number.isInteger(value);
            case DataType.FLOAT:
                return typeof value === 'number';
            case DataType.BOOLEAN:
                return typeof value === 'boolean';
            case DataType.DATETIME:
            case DataType.DATE:
            case DataType.TIME:
                return value instanceof Date || !isNaN(Date.parse(value));
            case DataType.ARRAY:
                return Array.isArray(value);
            case DataType.OBJECT:
                return typeof value === 'object' && !Array.isArray(value);
            default:
                return true;
        }
    }

    _validateConstraints(value) {
        // Numeric constraints
        if (typeof value === 'number') {
            if (this.minValue !== null && value < this.minValue) {
                throw new ValidationError(
                    `Field '${this.name}' value ${value} is below minimum ${this.minValue}`
                );
            }
            if (this.maxValue !== null && value > this.maxValue) {
                throw new ValidationError(
                    `Field '${this.name}' value ${value} exceeds maximum ${this.maxValue}`
                );
            }
        }

        // String constraints
        if (typeof value === 'string') {
            if (this.minLength !== null && value.length < this.minLength) {
                throw new ValidationError(
                    `Field '${this.name}' length ${value.length} is below minimum ${this.minLength}`
                );
            }
            if (this.maxLength !== null && value.length > this.maxLength) {
                throw new ValidationError(
                    `Field '${this.name}' length ${value.length} exceeds maximum ${this.maxLength}`
                );
            }
            if (this.pattern && !new RegExp(this.pattern).test(value)) {
                throw new ValidationError(
                    `Field '${this.name}' value does not match pattern ${this.pattern}`
                );
            }
        }

        // Enum constraint
        if (this.enum && !this.enum.includes(value)) {
            throw new ValidationError(
                `Field '${this.name}' value must be one of: ${this.enum.join(', ')}`
            );
        }
    }

    toJSON() {
        return {
            name: this.name,
            type: this.type,
            nullable: this.nullable,
            unique: this.unique,
            primaryKey: this.primaryKey,
            foreignKey: this.foreignKey,
            defaultValue: this.defaultValue,
            constraints: {
                minValue: this.minValue,
                maxValue: this.maxValue,
                minLength: this.minLength,
                maxLength: this.maxLength,
                pattern: this.pattern,
                enum: this.enum
            },
            description: this.description
        };
    }
}

/**
 * Represents a table or collection schema
 */
export class TableSchema {
    constructor(name, fields = [], options = {}) {
        this.name = name;
        this.fields = new Map();
        this.primaryKeys = [];
        this.foreignKeys = [];
        this.indexes = options.indexes || [];
        this.constraints = options.constraints || [];
        this.description = options.description || '';

        // Add fields
        for (const field of fields) {
            this.addField(field);
        }
    }

    addField(field) {
        if (!(field instanceof SchemaField)) {
            field = new SchemaField(field.name, field.type, field);
        }

        this.fields.set(field.name, field);

        if (field.primaryKey) {
            this.primaryKeys.push(field.name);
        }

        if (field.foreignKey) {
            this.foreignKeys.push({
                field: field.name,
                reference: field.foreignKey
            });
        }
    }

    removeField(name) {
        this.fields.delete(name);
        this.primaryKeys = this.primaryKeys.filter(pk => pk !== name);
        this.foreignKeys = this.foreignKeys.filter(fk => fk.field !== name);
    }

    getField(name) {
        return this.fields.get(name);
    }

    hasField(name) {
        return this.fields.has(name);
    }

    validate(record) {
        const errors = [];

        // Check for required fields
        for (const [name, field] of this.fields) {
            if (!field.nullable && !(name in record)) {
                errors.push(`Missing required field: ${name}`);
                continue;
            }

            if (name in record) {
                try {
                    field.validate(record[name]);
                } catch (error) {
                    errors.push(error.message);
                }
            }
        }

        if (errors.length > 0) {
            throw new SchemaError(
                `Schema validation failed: ${errors.join('; ')}`,
                this.name
            );
        }

        return true;
    }

    toJSON() {
        return {
            name: this.name,
            fields: Array.from(this.fields.values()).map(f => f.toJSON()),
            primaryKeys: this.primaryKeys,
            foreignKeys: this.foreignKeys,
            indexes: this.indexes,
            constraints: this.constraints,
            description: this.description
        };
    }

    static fromJSON(json) {
        const fields = json.fields.map(f => 
            new SchemaField(f.name, f.type, f)
        );
        
        return new TableSchema(json.name, fields, {
            indexes: json.indexes,
            constraints: json.constraints,
            description: json.description
        });
    }
}

/**
 * Schema detector for automatic schema inference
 */
export class SchemaDetector {
    constructor(options = {}) {
        this.sampleSize = options.sampleSize || 100;
        this.strictMode = options.strictMode || false;
        this.dateFormats = options.dateFormats || [
            /^\d{4}-\d{2}-\d{2}$/,
            /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/,
            /^\d{2}\/\d{2}\/\d{4}$/
        ];
    }

    detectFromSample(data) {
        if (!Array.isArray(data) || data.length === 0) {
            throw new SchemaError('Cannot detect schema from empty or non-array data');
        }

        const sample = data.slice(0, this.sampleSize);
        const fieldStats = this._analyzeFields(sample);
        const fields = this._inferFields(fieldStats);

        return new TableSchema('detected', fields);
    }

    _analyzeFields(sample) {
        const fieldStats = new Map();

        for (const record of sample) {
            if (typeof record !== 'object' || record === null) continue;

            for (const [key, value] of Object.entries(record)) {
                if (!fieldStats.has(key)) {
                    fieldStats.set(key, {
                        name: key,
                        types: new Map(),
                        nullCount: 0,
                        uniqueValues: new Set(),
                        minValue: null,
                        maxValue: null,
                        minLength: null,
                        maxLength: null
                    });
                }

                const stats = fieldStats.get(key);

                if (value === null || value === undefined) {
                    stats.nullCount++;
                } else {
                    const type = this._detectType(value);
                    stats.types.set(type, (stats.types.get(type) || 0) + 1);
                    stats.uniqueValues.add(JSON.stringify(value));

                    // Update numeric stats
                    if (typeof value === 'number') {
                        stats.minValue = stats.minValue === null ? value : Math.min(stats.minValue, value);
                        stats.maxValue = stats.maxValue === null ? value : Math.max(stats.maxValue, value);
                    }

                    // Update string stats
                    if (typeof value === 'string') {
                        const len = value.length;
                        stats.minLength = stats.minLength === null ? len : Math.min(stats.minLength, len);
                        stats.maxLength = stats.maxLength === null ? len : Math.max(stats.maxLength, len);
                    }
                }
            }
        }

        return fieldStats;
    }

    _detectType(value) {
        if (value === null || value === undefined) {
            return DataType.NULL;
        }

        if (typeof value === 'boolean') {
            return DataType.BOOLEAN;
        }

        if (typeof value === 'number') {
            return Number.isInteger(value) ? DataType.INTEGER : DataType.FLOAT;
        }

        if (typeof value === 'string') {
            // Check for date/time patterns
            for (const pattern of this.dateFormats) {
                if (pattern.test(value)) {
                    if (value.includes('T') || value.includes(' ')) {
                        return DataType.DATETIME;
                    }
                    return DataType.DATE;
                }
            }

            // Check for numeric strings
            if (/^-?\d+$/.test(value)) {
                return DataType.INTEGER;
            }
            if (/^-?\d+\.\d+$/.test(value)) {
                return DataType.FLOAT;
            }
            if (/^(true|false)$/i.test(value)) {
                return DataType.BOOLEAN;
            }

            return DataType.STRING;
        }

        if (Array.isArray(value)) {
            return DataType.ARRAY;
        }

        if (typeof value === 'object') {
            return DataType.OBJECT;
        }

        return DataType.UNKNOWN;
    }

    _inferFields(fieldStats) {
        const fields = [];

        for (const [name, stats] of fieldStats) {
            // Determine primary type
            let primaryType = DataType.STRING;
            let maxTypeCount = 0;

            for (const [type, count] of stats.types) {
                if (count > maxTypeCount) {
                    maxTypeCount = count;
                    primaryType = type;
                }
            }

            // Create field with inferred properties
            const field = new SchemaField(name, primaryType, {
                nullable: stats.nullCount > 0,
                unique: stats.uniqueValues.size === (stats.types.size + stats.nullCount),
                minValue: stats.minValue,
                maxValue: stats.maxValue,
                minLength: stats.minLength,
                maxLength: stats.maxLength
            });

            fields.push(field);
        }

        return fields;
    }
}

/**
 * Schema validator
 */
export class SchemaValidator extends BaseValidator {
    constructor(schema, strict = false) {
        super(strict);
        this.schema = schema instanceof TableSchema ? schema : TableSchema.fromJSON(schema);
    }

    validate(record) {
        try {
            return this.schema.validate(record);
        } catch (error) {
            this.addError(error.message);
            return false;
        }
    }
}

/**
 * Utility function to detect schema from data
 */
export function detectSchema(data, options = {}) {
    const detector = new SchemaDetector(options);
    return detector.detectFromSample(data);
}

/**
 * Utility function to validate data against schema
 */
export function validateSchema(data, schema) {
    const validator = new SchemaValidator(schema);
    
    if (Array.isArray(data)) {
        const errors = [];
        for (let i = 0; i < data.length; i++) {
            if (!validator.validate(data[i])) {
                errors.push(`Record ${i}: ${validator.getErrors().join(', ')}`);
                validator.clearErrors();
            }
        }
        if (errors.length > 0) {
            throw new SchemaError(`Validation failed: ${errors.join('; ')}`);
        }
        return true;
    } else {
        return validator.validate(data);
    }
}

/**
 * Convert schema between formats
 */
export function convertSchema(schema, targetFormat) {
    // Implement schema conversion logic for different formats
    // e.g., JSON Schema, Avro, Parquet, etc.
    switch (targetFormat) {
        case 'jsonschema':
            return convertToJSONSchema(schema);
        case 'sql':
            return convertToSQL(schema);
        default:
            throw new SchemaError(`Unsupported target format: ${targetFormat}`);
    }
}

function convertToJSONSchema(schema) {
    const jsonSchema = {
        type: 'object',
        properties: {},
        required: []
    };

    for (const [name, field] of schema.fields) {
        const property = {
            type: mapTypeToJSONSchema(field.type)
        };

        if (field.description) {
            property.description = field.description;
        }

        if (field.enum) {
            property.enum = field.enum;
        }

        if (field.pattern) {
            property.pattern = field.pattern;
        }

        jsonSchema.properties[name] = property;

        if (!field.nullable) {
            jsonSchema.required.push(name);
        }
    }

    return jsonSchema;
}

function mapTypeToJSONSchema(type) {
    const typeMap = {
        [DataType.STRING]: 'string',
        [DataType.INTEGER]: 'integer',
        [DataType.FLOAT]: 'number',
        [DataType.BOOLEAN]: 'boolean',
        [DataType.DATETIME]: 'string',
        [DataType.DATE]: 'string',
        [DataType.TIME]: 'string',
        [DataType.ARRAY]: 'array',
        [DataType.OBJECT]: 'object',
        [DataType.NULL]: 'null'
    };
    return typeMap[type] || 'string';
}

function convertToSQL(schema) {
    const columns = [];
    
    for (const [name, field] of schema.fields) {
        let columnDef = `${name} ${mapTypeToSQL(field.type)}`;
        
        if (!field.nullable) {
            columnDef += ' NOT NULL';
        }
        
        if (field.primaryKey) {
            columnDef += ' PRIMARY KEY';
        }
        
        if (field.unique) {
            columnDef += ' UNIQUE';
        }
        
        if (field.defaultValue !== null) {
            columnDef += ` DEFAULT ${field.defaultValue}`;
        }
        
        columns.push(columnDef);
    }
    
    return `CREATE TABLE ${schema.name} (\n  ${columns.join(',\n  ')}\n);`;
}

function mapTypeToSQL(type) {
    const typeMap = {
        [DataType.STRING]: 'VARCHAR(255)',
        [DataType.INTEGER]: 'INTEGER',
        [DataType.FLOAT]: 'FLOAT',
        [DataType.BOOLEAN]: 'BOOLEAN',
        [DataType.DATETIME]: 'TIMESTAMP',
        [DataType.DATE]: 'DATE',
        [DataType.TIME]: 'TIME',
        [DataType.ARRAY]: 'JSON',
        [DataType.OBJECT]: 'JSON',
        [DataType.NULL]: 'VARCHAR(255)'
    };
    return typeMap[type] || 'VARCHAR(255)';
}